### PR TITLE
Adding interface area to two phase

### LINKED ISSUE #27444: Adding interface area tracking to two-phase mixture model
### Motivation

For modeling two-phase interface phenomena, the specific interface area between the liquid and the dispersed phases need to be tracked. This is currenty not included in Navier Stoke's mixture models.

### Design

- The model should follow a multi-D generalization of the commonly used 1D model developed by Hibiki and Ishii in this [article](https://www.sciencedirect.com/science/article/pii/S0017931099003257).
- The model should work with the mixture formulation for two-phase modeling in the MOOSE's Navier Stokes module.

### Impact

The addition of this model provides a wide range of functionality for two-phase mass transfer modeling, including species tracking in MSRs or two-phase fluidized bed separations.

### PR DESCRIPTION
**Pull request by tanoret**
Closes #27444

### CODE REVIEW

**Comment by tanoret** [comment]
@MengnanLi91 @grmnptr please take a look at your convenience.

The model has been validated against the experimental results in [Hibiki and Ishii](https://www.sciencedirect.com/science/article/pii/S0017931099003257).

A simple version of two of the experiments in that paper is included in the test regression suite.

Thank you very much!

File: modules/navier_stokes/doc/content/source/fvkernels/WCNSFV2PInterfaceAreaSourceSink.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+-\frac{1}{3} \frac{D \rho_d}{Dt} + S_T + \rho_d (S_C + S_B)\,,
+\end{equation}
+
+where:
+
+- $\rho_d$ is the density of the dispersed phase $d$, e.g., the density of the gas in bubbles,
+- $t$ is time,
+- $\vec{u}$ is the velocity vector,
+- $D_b$ is a diffusion coefficient for the interface area concentration, which may be assumed as `0` if unknown,
+- $\frac{D (\bullet)}{Dt} = \frac{\partial (\bullet)}{\partial t} + \vec{u} \cdot \nabla (\bullet)$ is the material derivative,
```
Comment (grmnptr) [inline modules/navier_stokes/doc/content/source/fvkernels/WCNSFV2PInterfaceAreaSourceSink.md:20]: If you are using an advection kernel we already have then it is using `div(u*phi)` instead of `u*grad(phi)`. The two are only the same (assuming that they have the partial with time as well) if we have a specific quantity (so we conserve rho*phi) and we respect the continuity with u. Not sure if this matters in this case, feel free to disregard this if this comment is irrevelant.

File: modules/navier_stokes/doc/content/source/fvkernels/WCNSFV2PInterfaceAreaSourceSink.md [hunk 2]
```diff
@@ -0,0 +1,114 @@
+# WCNSFV2PInterfaceAreaSourceSink
```
Comment (grmnptr) [inline modules/navier_stokes/doc/content/source/fvkernels/WCNSFV2PInterfaceAreaSourceSink.md:1]: Very nice documentation, you should publish this in a paper as well!

File: modules/navier_stokes/doc/content/source/fvkernels/WCNSFV2PInterfaceAreaSourceSink.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+- $\tilde{d_p} = \psi \frac{\alpha_d}{\xi_p}$ is the model estimate for the dispersed phase particle diameter, with $\psi$ being a shape factor, which is, for example, $\psi=6$ for spherical particles,
+- $\alpha_d^{max}$ is the maximum volumetric fraction admitted by the model, in absence of data we recommend taking $\alpha_d^{max}=1$,
+- $\rho_b$ is the bulk phase density, e.g., for air bubbles in a water flow it would be the density of water,
+- $\sigma$ is the surface tension between the two phases,
+- $\Gamma_C = 0.188$ and $K_C = 0.129$ are closure coefficients of the model.
+
+!alert note
+Many of the parameters in the coalescence model are provided as functors,
+which means that spatially dependent fields may be specified for these parameters.
+However, the validation of this model has only been performed using constant parameters for this model.
```
Comment (grmnptr) [inline modules/navier_stokes/doc/content/source/fvkernels/WCNSFV2PInterfaceAreaSourceSink.md:83]:
```suggestion
However, the validation of this model has only been performed using constant parameters.
```

File: modules/navier_stokes/doc/content/source/fvkernels/WCNSFV2PInterfaceAreaSourceSink.md [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+- $\tilde{d_p} = \psi \frac{\alpha_d}{\xi_p}$ is the model estimate for the dispersed phase particle diameter, with $\psi$ being a shape factor, which is, for example, $\psi=6$ for spherical particles,
+- $\alpha_d^{max}$ is the maximum volumetric fraction admitted by the model, in absence of data we recommend taking $\alpha_d^{max}=1$,
+- $\rho_b$ is the bulk phase density, e.g., for air bubbles in a water flow it would be the density of water,
+- $\sigma$ is the surface tension between the two phases,
+- $\Gamma_B = 0.264$ and $K_B = 1.370$ are closure coefficients of the model.
+
+!alert note
+Many of the parameters in the breakage model are provided as functors,
+which means that spatially dependent fields may be specified for these parameters.
+However, the validation of this model has only been performed using constant parameters for this model.
```
Comment (grmnptr) [inline modules/navier_stokes/doc/content/source/fvkernels/WCNSFV2PInterfaceAreaSourceSink.md:107]:
```suggestion
However, the validation of this model has only been performed using constant parameters.
```

File: modules/navier_stokes/include/fvkernels/WCNSFV2PInterfaceAreaSourceSink.h [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "FVElementalKernel.h"
+#include "INSFVVelocityVariable.h"
+
+/**
+ * Computes source the sink terms for the turbulent kinetic energy.
```
Comment (grmnptr) [inline modules/navier_stokes/include/fvkernels/WCNSFV2PInterfaceAreaSourceSink.h:16]:
```suggestion
 * Computes source the sink terms for the interface area in the mixture model of two-phase flows.
```

File: modules/navier_stokes/include/fvkernels/WCNSFV2PInterfaceAreaSourceSink.h [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+  const Moose::Functor<ADReal> & _pressure;
+
+  /// Interface Mass Exchange Coeefficient
+  const Moose::Functor<ADReal> & _mass_exchange_coefficient;
+
+  /// Void Fraction
+  const Moose::Functor<ADReal> & _f_d;
+
+  /// Maximum Void Fraction
+  const Real & _f_d_max;
```
Comment (grmnptr) [inline modules/navier_stokes/include/fvkernels/WCNSFV2PInterfaceAreaSourceSink.h:58]:
```suggestion
  const Real  _f_d_max;
```

Comment (grmnptr) [reply modules/navier_stokes/include/fvkernels/WCNSFV2PInterfaceAreaSourceSink.h:58]: Real can be more expensive to fetch through a reference. The memory consumption of a Real and a reference is the same.

File: modules/navier_stokes/include/fvkernels/WCNSFV2PInterfaceAreaSourceSink.h [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+  /// Cutoff fraction at which the full mass transfer model is activated
+  const Real _cutoff_fraction;
+
+  /// Internal closure coefficients
+  Real _gamma_c;
+  Real _Kc;
+  Real _gamma_b;
+  Real _Kb;
+  Real _shape_factor;
```
Comment (grmnptr) [inline modules/navier_stokes/include/fvkernels/WCNSFV2PInterfaceAreaSourceSink.h:74]: If you don't want these to be tunable then make them const. Or constexpr if you want the compiler to bake them in.

File: modules/navier_stokes/src/fvkernels/WCNSFV2PInterfaceAreaSourceSink.C [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+  params.addRequiredParam<MooseFunctorName>(NS::density + std::string("_d"),
+                                            "Dispersed phase density.");
+  params.addRequiredParam<MooseFunctorName>(NS::pressure, "Continuous phase density.");
+  params.addParam<MooseFunctorName>(
+      "k_c", 0.0, "Mass exchange coefficients from continous to dispersed phases.");
+  params.addParam<MooseFunctorName>("fd", 0.0, "Fraction dispersed phase.");
+  params.addParam<Real>("fd_max", 1.0, "Maximum dispersed phase fraction.");
+
+  params.addParam<MooseFunctorName>("sigma", 1.0, "Surface tension between phases.");
+  params.addParam<MooseFunctorName>("particle_diameter", 1.0, "Maximum Particle Diameter.");
```
Comment (grmnptr) [inline modules/navier_stokes/src/fvkernels/WCNSFV2PInterfaceAreaSourceSink.C:38]:
```suggestion
  params.addParam<MooseFunctorName>("particle_diameter", 1.0, "Maximum particle diameter.");
```

File: modules/navier_stokes/doc/content/source/fvkernels/WCNSFV2PInterfaceAreaSourceSink.md (hunk 1 above)
Comment (grmnptr) [reply modules/navier_stokes/doc/content/source/fvkernels/WCNSFV2PInterfaceAreaSourceSink.md:20]: Okay, I see the code, it is consistent, so disregard the comment above.

File: modules/navier_stokes/src/fvkernels/WCNSFV2PInterfaceAreaSourceSink.C [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+  ADRealVectorValue velocity(u);
+  if (_v_var)
+    velocity(1) = (*_v_var)(elem_arg, state);
+  if (_w_var)
+    velocity(2) = (*_w_var)(elem_arg, state);
+
+  const ADReal velocity_norm = NS::computeSpeed(velocity);
+  const auto pressure_gradient = raw_value(_pressure.gradient(elem_arg, state));
+  const ADReal pressure_grad_norm =
+      MooseUtils::isZero(pressure_gradient) ? 1e-42 : pressure_gradient.norm();
```
Comment (grmnptr) [inline modules/navier_stokes/src/fvkernels/WCNSFV2PInterfaceAreaSourceSink.C:138]: If this discards derivatives, why not just real for the grad_norm? Also this might mess up the Jacobian if you are solving with the monolithic solver. Can you just lag the pressure gradient (so evaluate with `old_state`).

File: modules/navier_stokes/test/tests/finite_volume/two_phase/mixture_interface_area_model/tests [hunk 10]
```diff
@@ -0,0 +1,28 @@
+[Tests]
+  issues = '#27444'
+  design = 'WCNSFV2PInterfaceAreaSourceSink.md'
+  [mixture_interface_pressure_driven]
+    type = 'Exodiff'
+    input = 'pressure_driven_growth.i'
+    exodiff = 'pressure_driven_growth_out.e'
+    rel_err = 1.0E-6
+    requirement = 'The system shall be able to track interfacial area in steady-state, two-phase problems dominated by pressure driven growth.'
+    installation_type = in_tree
```
Comment (grmnptr) [inline modules/navier_stokes/test/tests/finite_volume/two_phase/mixture_interface_area_model/tests:10]: Why does this have to be in tree?

File: modules/navier_stokes/test/tests/finite_volume/two_phase/mixture_interface_area_model/tests [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+  []
+  [mixture_interface_turbulent_driven]
+    type = 'Exodiff'
+    input = 'turbulent_driven_growth.i'
+    exodiff = 'turbulent_driven_growth_out.e'
+    rel_err = 1.0E-6
+    requirement = 'The system shall be able to track interfacial area in steady-state, two-phase problems dominated by turbulent driven growth.'
```
Comment (grmnptr) [inline modules/navier_stokes/test/tests/finite_volume/two_phase/mixture_interface_area_model/tests:17]:
```suggestion
    requirement = 'The system shall be able to track interfacial area in steady-state, two-phase problems dominated by turbulence driven growth.'
```

File: modules/navier_stokes/test/tests/finite_volume/two_phase/mixture_model/channel-drift-flux-w-interface-area.i [hunk 12]
```diff
@@ -0,0 +1,368 @@
+mu = 10.0
```
Comment (grmnptr) [inline modules/navier_stokes/test/tests/finite_volume/two_phase/mixture_model/channel-drift-flux-w-interface-area.i:1]: We need a test for this input file. Something very simple (decrease mesh resolution) to make sure it remains runnable no matter who changes what.

**Review by grmnptr — CHANGES_REQUESTED** [review]
Cool stuff, you should publish it!

File: modules/navier_stokes/doc/content/source/fvkernels/WCNSFV2PInterfaceAreaSourceSink.md [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+
+## Interface area concentration source due to breakage
+
+The increase in interface area concentration due to breakage of the dispersed phase is modeled as follows:
+
+\begin{equation}
+S_B = \left( \frac{\alpha_d}{\xi_p} \right)^2 \frac{\Gamma_B \alpha_d (1 - \alpha_d) u_{\epsilon}}{\tilde{d_p}^{11/3} (\alpha_d^{max}- \alpha_d)}
+\operatorname{exp} \left( -K_B \frac{\sigma}{\rho_b \tilde{d_p}^{5/6} u_{\epsilon}^{2/3}} \right)\,,
+\end{equation}
+
```
Comment (MengnanLi91) [inline modules/navier_stokes/doc/content/source/fvkernels/WCNSFV2PInterfaceAreaSourceSink.md:93]: Inside the exponential term, it should be $\tilde{d_p}^{5/3}$ instead of $\tilde{d_p}^{5/6}$ according to the reference paper. I checked, the code implementation is correct

File: modules/navier_stokes/doc/content/source/fvkernels/WCNSFV2PInterfaceAreaSourceSink.md [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+## Interface area concentration source due to coalescence
+
+The reduction in interface area concentration due to coalescence of the dispersed phase is modeled as follows:
+
+\begin{equation}
+S_C = -\left( \frac{\alpha_d}{\xi_p} \right)^2 \frac{\Gamma_C \alpha_d^2 u_{\epsilon}}{\tilde{d_p}^{11/3} (\alpha_d^{max}- \alpha_d)}
+\operatorname{exp} \left( -K_C \frac{\tilde{d_p}^{5/6} \rho_b^{1/2} u_{\epsilon}^{1/3}}{\sigma^{1/2}} \right)\,,
+\end{equation}
+
```
[the remark this answers is not in the document]
Comment (tanoret) [reply modules/navier_stokes/doc/content/source/fvkernels/WCNSFV2PInterfaceAreaSourceSink.md:69]: Thanks, I think this one was already added to the main equation in the term $\rho_d (S_C + S_B)$, would you please confirm? Is other the density that is missing?

File: modules/navier_stokes/doc/content/source/fvkernels/WCNSFV2PInterfaceAreaSourceSink.md [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+\frac{\partial (\rho_d \xi_p)}{\partial t} + \nabla \cdot \left( \rho_d \vec{u} \xi_p \right) - \nabla \left( D_b \nabla \xi_p \right) =
+-\frac{1}{3} \frac{D \rho_d}{Dt} + S_T + \rho_d (S_C + S_B)\,,
+\end{equation}
+
+where:
+
+- $\rho_d$ is the density of the dispersed phase $d$, e.g., the density of the gas in bubbles,
+- $t$ is time,
+- $\vec{u}$ is the velocity vector,
+- $D_b$ is a diffusion coefficient for the interface area concentration, which may be assumed as `0` if unknown,
```
[the remark this answers is not in the document]
Comment (tanoret) [reply modules/navier_stokes/doc/content/source/fvkernels/WCNSFV2PInterfaceAreaSourceSink.md]: Thank you. It would depend on the units you assign to the diffusion coefficient. In this case it is a mass diffusion coefficient. The units should be $\frac{kg}{m.s}$. We could multiply by the density to get a standard mollecular diffusion coefficient. However, since this is more of a tunable parameter, I don't think it's worth it. What do you think?

Regarding the reference, the interfacial tracking models were developed for 1D, which does not account for cross-streamline diffusion. We are the first team trying to generalize them to multi-D my knowledge. We need cross streamline in multi-D models to get the correct radial void profile. I think the description that says "which may be assumed to be 0 if unknown" is enough description to avoid misuse. What do you think?

File: modules/navier_stokes/doc/content/source/fvkernels/WCNSFV2PInterfaceAreaSourceSink.md (hunk 1 above)
Comment (tanoret) [reply modules/navier_stokes/doc/content/source/fvkernels/WCNSFV2PInterfaceAreaSourceSink.md:20]: Thanks. Resolving

File: modules/navier_stokes/doc/content/source/fvkernels/WCNSFV2PInterfaceAreaSourceSink.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (MengnanLi91) [reply modules/navier_stokes/doc/content/source/fvkernels/WCNSFV2PInterfaceAreaSourceSink.md:69]: Yes, you are right!

File: modules/navier_stokes/doc/content/source/fvkernels/WCNSFV2PInterfaceAreaSourceSink.md [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+two phases per unit volume, i.e.,$[\xi_p] = \frac{m^2}{m^3}$. This parameter is important
+for predicting mass, momentum, and energy transfer at interfaces in two-phase flows.
+
+The general equation for interfacial area concentration transport via the mixture model reads as follows:
+
+\begin{equation}
+\frac{\partial (\rho_d \xi_p)}{\partial t} + \nabla \cdot \left( \rho_d \vec{u} \xi_p \right) - \nabla \left( D_b \nabla \xi_p \right) =
+-\frac{1}{3} \frac{D \rho_d}{Dt} + S_T + \rho_d (S_C + S_B)\,,
+\end{equation}
+
```
Comment (MengnanLi91) [inline modules/navier_stokes/doc/content/source/fvkernels/WCNSFV2PInterfaceAreaSourceSink.md:13]: I didn't mark the right place for density term missing. Here is what I mean, the diffusion term here has a unit discrepancy.

File: modules/navier_stokes/include/fvkernels/WCNSFV2PInterfaceAreaSourceSink.h (hunk 7 above)
Comment (tanoret) [reply modules/navier_stokes/include/fvkernels/WCNSFV2PInterfaceAreaSourceSink.h]: Changed to `constexpr`

File: modules/navier_stokes/doc/content/source/fvkernels/WCNSFV2PInterfaceAreaSourceSink.md (hunk 2 above)
Comment (tanoret) [reply modules/navier_stokes/doc/content/source/fvkernels/WCNSFV2PInterfaceAreaSourceSink.md:1]: Thanks

File: modules/navier_stokes/src/fvkernels/WCNSFV2PInterfaceAreaSourceSink.C (hunk 9 above)
Comment (tanoret) [reply modules/navier_stokes/src/fvkernels/WCNSFV2PInterfaceAreaSourceSink.C]: Chaged according to oral discussion.

File: modules/navier_stokes/test/tests/finite_volume/two_phase/mixture_interface_area_model/tests (hunk 10 above)
Comment (tanoret) [reply modules/navier_stokes/test/tests/finite_volume/two_phase/mixture_interface_area_model/tests]: Removed the in_tree testing

File: modules/navier_stokes/test/tests/finite_volume/two_phase/mixture_model/channel-drift-flux-w-interface-area.i (hunk 12 above)
Comment (tanoret) [reply modules/navier_stokes/test/tests/finite_volume/two_phase/mixture_model/channel-drift-flux-w-interface-area.i:1]: Added test for input file.

File: modules/navier_stokes/doc/content/source/fvkernels/WCNSFV2PInterfaceAreaSourceSink.md (hunk 13 above)
Comment (tanoret) [reply modules/navier_stokes/doc/content/source/fvkernels/WCNSFV2PInterfaceAreaSourceSink.md:93]: Thanks. Corrected.

File: modules/navier_stokes/doc/content/source/fvkernels/WCNSFV2PInterfaceAreaSourceSink.md (hunk 16 above)
Comment (MengnanLi91) [reply modules/navier_stokes/doc/content/source/fvkernels/WCNSFV2PInterfaceAreaSourceSink.md:13]: Ok, I thought the unit is m2/s. Then it should be fine.

File: modules/navier_stokes/test/tests/finite_volume/two_phase/mixture_model/channel-drift-flux-w-interface-area.i
Comment (MengnanLi91) [inline modules/navier_stokes/test/tests/finite_volume/two_phase/mixture_model/channel-drift-flux-w-interface-area.i:1]: Is this the test case validated against [Hibiki and Ishii](https://www.sciencedirect.com/science/article/pii/S0017931099003257)? Which quantity is compared? The flow properties and geometry seem very different from experiment.

Comment (tanoret) [reply modules/navier_stokes/test/tests/finite_volume/two_phase/mixture_model/channel-drift-flux-w-interface-area.i:1]: No. This is just a non-regression test. I may come back to a few of these tests later today as a few of them are failing...

Comment (tanoret) [reply modules/navier_stokes/test/tests/finite_volume/two_phase/mixture_model/channel-drift-flux-w-interface-area.i:1]: For the validation tests, let's try to add the to the [flow test bed](https://github.com/neams-th-coe/flow_test_bed), which is currently quite unpopulated but should be growing soon :)

**Comment by tanoret** [comment]
All good with this? I cannot find any comments left

**Comment by MengnanLi91** [comment]
> All good with this? I cannot find any comments left

Peter requested changes so he has to lift that before I can merge it

File: modules/navier_stokes/src/fvkernels/WCNSFV2PInterfaceAreaSourceSink.C [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+  // Adding area growth due to added mass
+  ADReal bubble_added_mass;
+  if (f_d < _cutoff_fraction)
+    bubble_added_mass = raw_value(_rho_d(elem_arg, state)) *
+                        (f_d * 6.0 / _particle_diameter(elem_arg, state) - _var(elem_arg, state));
+  else
+    bubble_added_mass = 2. / 3. * _mass_exchange_coefficient(elem_arg, state) *
+                        (1.0 / (f_d + libMesh::TOLERANCE) - 2.0);
+
+  // // Model parameters
```
Comment (grmnptr) [inline modules/navier_stokes/src/fvkernels/WCNSFV2PInterfaceAreaSourceSink.C:121]:
```suggestion
  // Model parameters
```

File: modules/navier_stokes/test/tests/finite_element/cns/step/step_kernels.i [hunk 18]
```diff
@@ -0,0 +1,120 @@
+# Navier-Stokes (or Euler) flow of an ideal gas over a step.
```
Comment (grmnptr) [inline modules/navier_stokes/test/tests/finite_element/cns/step/step_kernels.i:1]: Is this file related to this PR? It looks like it isn't.

**Review by grmnptr — CHANGES_REQUESTED** [review]
Almost there!

**Review by grmnptr — APPROVED** [review]
Nice!

**Review by GiudGiud — DISMISSED** [review]
the commit history is not useful / unclear.
"Addressing review" does not say what the commit does. 

and you are checking in regold-ed files twice (three times for one), which is the bigger issue.

**Comment by tanoret** [comment]
Gold files removed from previous commits

**Comment by GiudGiud** [comment]
Merging with Peter's prior approval
